multithreading/WebPagesDownloader/src/WebPagesDownloader/WebPagesDownloader.fs

```
@@ -0,0 +1,122 @@
+        /// Регулярное выражение для поиска ссылок на указанной странице.
+        let findInnerWebPages = new Regex(@"a href=""http://(\S*)""")
+        
+        /// Нормализовать найденный шаблон `a href="http://..."` до `http://...`
+        let normalizeUrl = new Regex(@"a href=|""")        
+
+        /// Скачать html.
+        let downloadHtml url = 
+            async {
+                do printfn "Creating request for %s..." url
+                try                
+                    let request = WebRequest.Create(url)
+                    use! response = request.AsyncGetResponse() 
+                    do printfn "Getting response stream for %s..." url
+                    use stream = response.GetResponseStream()
+                    if stream = null then 
+                        printf "Stream from response is null. Exit from downloading process."
+                        return "-1"
+                    else 
+                        do printfn "Reading response for %s..." url
+                        use reader = new StreamReader(stream)
+                        return reader.ReadToEnd()
+                with 
+                | :? System.Security.SecurityException as e ->
+                    printfn "%s" e.Message
+                    return "-1"
+                | :? System.Net.WebException as e -> 
+                    printfn "%s" e.Message
+                    return "-1"
+                | :? System.NotSupportedException as e ->
+                    printfn "%s" e.Message
+                    return "-1"
+                | :? System.ArgumentException as e ->
+                    printfn "%s" e.Message
+                    return "-1"
+                | :? System.OutOfMemoryException as e ->
+                    printfn "%s" e.Message
+                    return "-1"
+                | :? System.IO.IOException as e ->
+                    printfn "%s" e.Message
+                    return "-1"                
```

Ой. Вообще, есть хороший принцип "Fail Fast" --- не надо печатать исключения, надо дать программе упасть, они сами напечатаются. Быть готовым продолжить работу в случае неправильного ответа сервера хорошо, ловить OutOfMemoryException или ArgumentException плохо